[JSC] Optimize DataView.prototype.byteLength handling in JIT

hyjorc1 opened the pull request:
#### e6f31f64c264e758f2f357b83e71dd404d6c8f73
<pre>
[JSC] Optimize DataView.prototype.byteLength handling in JIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=289365">https://bugs.webkit.org/show_bug.cgi?id=289365</a>
<a href="https://rdar.apple.com/146511506">rdar://146511506</a>

Reviewed by Yusuke Suzuki.

Introduced DataViewByteLengthIntrinsic to improve inline caching and JIT
optimizations for DataView.prototype.byteLength. With this change, the
byteLength property’s order is rearranged in the DataView property table.
Consequently, object-expected.txt has been updated accordingly.

MicroBenchmarks Results:

                                           without                     with

dataview-byte-length-small-array       14.7764+-0.0405     ^      4.6604+-0.0182        ^ definitely 2.8164x faster

* JSTests/microbenchmarks/dataview-byte-length-small-array.js: Added.
(test):
* JSTests/stress/dataview-byte-length-large-array-oob.js: Added.
(test):
(i.i.testLoopCount.1.catch):
* JSTests/stress/dataview-byte-length-large-array.js: Added.
(test):
* JSTests/stress/dataview-byte-length-small-array-oob.js: Added.
(test):
(i.i.testLoopCount.1.catch):
* JSTests/stress/dataview-byte-length-small-array.js: Added.
(test):
* Source/JavaScriptCore/assembler/AbortReason.h:
* Source/JavaScriptCore/bytecode/GetByVariant.cpp:
(JSC::GetByVariant::canMergeIntrinsicStructures const):
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheCompiler::canEmitIntrinsicGetter):
(JSC::InlineCacheCompiler::emitIntrinsicGetter):
* Source/JavaScriptCore/bytecode/IntrinsicGetterAccessCase.cpp:
(JSC::IntrinsicGetterAccessCase::doesCalls const):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGArrayMode.cpp:
(JSC::DFG::ArrayMode::alreadyChecked const):
(JSC::DFG::toTypedArrayType):
(JSC::DFG::toArrayType):
* Source/JavaScriptCore/dfg/DFGArrayMode.h:
(JSC::DFG::ArrayMode::isSomeTypedArrayViewOrDataView const):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicGetter):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByVal):
(JSC::DFG::SpeculativeJIT::compileGetTypedArrayLengthAsInt52):
(JSC::DFG::SpeculativeJIT::compilePutByVal):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::typedArrayLength):
(JSC::FTL::DFG::LowerDFGToB3::compileGetArrayLength):
(JSC::FTL::DFG::LowerDFGToB3::compileGetTypedArrayLengthAsInt52):
(JSC::FTL::DFG::LowerDFGToB3::compileGetByValImpl):
(JSC::FTL::DFG::LowerDFGToB3::compilePutByVal):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::loadTypedArrayByteLengthImpl):
(JSC::AssemblyHelpers::loadTypedArrayByteLengthCommonImpl):
(JSC::AssemblyHelpers::loadTypedArrayByteLength):
(JSC::AssemblyHelpers::loadDataViewByteLength):
(JSC::AssemblyHelpers::loadTypedArrayLength):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::loadDataViewByteLength):
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/jit/JITOperations.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
(JSC::JSDataViewPrototype::create):
(JSC::JSDataViewPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSDataViewPrototype.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):

Canonical link: <a href="https://commits.webkit.org/291993@main">https://commits.webkit.org/291993@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/165a7158cb69e9d3fd9b9df27630d2aeb50f2120

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94560 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14152 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3951 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99580 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45075 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14450 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22580 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72163 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29476 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97562 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10760 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85379 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52494 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10454 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3081 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44399 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/87241 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80670 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3187 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101623 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/93207 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21616 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15772 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81162 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21866 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81402 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80537 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20108 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25092 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2476 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14835 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21593 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26722 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/115874 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21272 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24735 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23010 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -4240,10 +4240,12 @@ bool InlineCacheCompiler::canEmitIntrinsicGetter(StructureStubInfo& stubInfo, JS
         return false;
 
     switch (getter->intrinsic()) {
+    case DataViewByteLengthIntrinsic:
```
Maybe separating clauses would make things easier to add other DataView function supports.
```
case DataViewByteLengthIntrinsic: {
    if (structure->typeInfo().type() != DataViewType)
        return false;
#if USE(JSVALUE32_64)
    if (isResizableOrGrowableSharedTypedArrayIncludingDataView(structure->classInfoForCells()))
        return false;
#endif
    return true;
}
```

Constellation commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -4319,7 +4322,30 @@ void InlineCacheCompiler::emitIntrinsicGetter(IntrinsicGetterAccessCase& accessC
+                jit.reclaimSpaceOnStackForCCall();
+
+                restoreLiveRegistersFromStackForCallWithThrownException(spillState);
+                emitExplicitExceptionHandler();
```
Let's make IC code more simple and tight. I think you can just jump to `m_failAndIgnore` at the end.
You also need to do `allocator.restoreReusedRegistersByPopping` before that jump. And these code generation should be done after `succeed()` call to make hot path code tight.
See `AccessCase::IndexedInt32Load`'s `m_failAndIgnore` jump code for example.

Constellation commented on Source/JavaScriptCore/dfg/DFGArrayMode.cpp:
```diff
@@ -661,6 +661,9 @@ bool ArrayMode::alreadyChecked(Graph& graph, Node* node, const AbstractValue& va
         return speculationChecked(value.m_type, SpecTypedArrayView);
 
+    case Array::DataView:
+        return speculationChecked(value.m_type, SpecDataViewObject);
```
Since `DataView` is not array, I don't think we should extend ArrayMode. Just use DataViewUse speculation and that's fine.

Constellation commented on Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
```diff
@@ -417,7 +417,7 @@ class ConstantFoldingPhase : public Phase {
                 ArrayMode arrayMode = node->arrayMode();
                 AbstractValue& abstractValue = m_state.forNode(node->child1());
-                if (arrayMode.type() != Array::AnyTypedArray && arrayMode.isSomeTypedArrayView() && !arrayMode.mayBeResizableOrGrowableSharedTypedArray()) {
+                if (arrayMode.type() != Array::AnyTypedArray && arrayMode.isSomeTypedArrayViewOrDataView() && !arrayMode.mayBeResizableOrGrowableSharedTypedArray()) {
```
I think this should not include DataView since `DataView.length` does not exist.

Constellation commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -2428,7 +2428,7 @@ class FixupPhase : public Phase {
             if (arrayMode.type() == Array::Generic)
                 arrayMode = arrayMode.withType(Array::ForceExit);
-            DFG_ASSERT(m_graph, node, arrayMode.isSomeTypedArrayView() || arrayMode.type() == Array::ForceExit);
+            DFG_ASSERT(m_graph, node, arrayMode.isSomeTypedArrayViewOrDataView() || arrayMode.type() == Array::ForceExit);
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/JITOperations.cpp:
```diff
@@ -4601,6 +4601,17 @@ JSC_DEFINE_JIT_OPERATION(operationReallocateButterflyToGrowPropertyStorage, char
+    throwVMTypeError(globalObject, scope, typedArrayBufferHasBeenDetachedErrorMessage);
+    OPERATION_RETURN(scope);
+}
+
```
Not necessary if we just jump to slow C++ code

hyjorc1 replied:
I think we still need that for DFG and FTL, right? Since `CheckIsDetached` doesn't check OOB case.

Constellation commented on Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
```diff
@@ -234,12 +235,13 @@ JSC_DEFINE_CUSTOM_GETTER(dataViewProtoGetterBuffer, (JSGlobalObject* globalObjec
-    JSDataView* view = jsDynamicCast<JSDataView*>(JSValue::decode(thisValue));
+    JSValue thisValue = callFrame->thisValue();
+    JSDataView* view = jsDynamicCast<JSDataView*>(thisValue);
     if (!view)
```
Let's put `UNLIKELY`

Constellation commented on Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
```diff
@@ -234,12 +235,13 @@ JSC_DEFINE_CUSTOM_GETTER(dataViewProtoGetterBuffer, (JSGlobalObject* globalObjec
 
-    JSDataView* view = jsDynamicCast<JSDataView*>(JSValue::decode(thisValue));
+    JSValue thisValue = callFrame->thisValue();
+    JSDataView* view = jsDynamicCast<JSDataView*>(thisValue);
```
You can just do,
```
JSDataView* view = jsDynamicCast<JSDataView*>(callFrame->thisValue());
```

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -4375,6 +4375,7 @@ bool ByteCodeParser::handleIntrinsicGetter(Operand result, SpeculatedType predic
         return false;
 
     switch (variant.intrinsic()) {
+    case DataViewByteLengthIntrinsic:
```
For `DataViewByteLengthIntrinsic` case, you can emit `CheckDetached` to ensure that DataView is not detached. (And then you can also ensure that the array is not detached in code generation).
And you can stop inlining here when we had `BadIndexingType` OSR exit kind here via `hasExitSite`.

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -4375,6 +4375,7 @@ bool ByteCodeParser::handleIntrinsicGetter(Operand result, SpeculatedType predic
 
     switch (variant.intrinsic()) {
+    case DataViewByteLengthIntrinsic:
     case TypedArrayByteLengthIntrinsic: {
```
Let's not reuse `ArrayMode` based code below since DataView is not Array. Let's add new DataViewGetByteLength and DataViewGetByteLengthAsInt52 node types and just use DataViewUse edge.

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -4376,10 +4376,37 @@ bool ByteCodeParser::handleIntrinsicGetter(Operand result, SpeculatedType predic
+        if (m_inlineStackTop->m_exitProfile.hasExitSite(m_currentIndex, BadIndexingType))
+            return false;
+
+        addToGraph(CheckDetached, Edge(thisNode, DataViewObjectUse));
```
Let's emit something after doing all the checks. So emitting it after `mayBeLargeArrayBuffer` check is done.

Also, CheckDetached is always emitting `CellUse` in the fixup for the child. So let's emit `DataViewObjectUse` check separately,

```
addToGraph(Check, Edge(thisNode, DataViewObjectUse));
addToGraph(CheckDetached, thisNode);
```

Constellation commented on Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:
```diff
@@ -1516,8 +1516,11 @@ class IntegerRangeOptimizationPhase : public Phase {
+        case DataViewGetByteLength:
+        case DataViewGetByteLengthAsInt52:
         case GetArrayLength:
+        case GetTypedArrayLengthAsInt52:
```
I think integer range optimization phase cannot handle Int52 well. So `GetTypedArrayLengthAsInt52` is intentionally removed from this switch. Let's remove `DataViewGetByteLengthAsInt52` too.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -8132,6 +8132,46 @@ void SpeculativeJIT::compileGetArrayLength(Node* node)
+        GPRReg baseGPR = base.gpr();
+        GPRReg scratch1GPR = scratch1.gpr();
+        GPRReg resultGPR = result.gpr();
+
```
Use here.
```
speculateDataViewObject(node->child1(), baseGPR);
```

SpeculateCellOperand only checks whether it is cell.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -8132,6 +8132,46 @@ void SpeculativeJIT::compileGetArrayLength(Node* node)
+    GPRTemporary result(this);
+    GPRReg baseGPR = base.gpr();
+    GPRReg resultGPR = result.gpr();
+
```
Ditto.
Use here.
```
speculateDataViewObject(node->child1(), baseGPR);
```

SpeculateCellOperand only checks whether it is cell.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -8132,6 +8132,46 @@ void SpeculativeJIT::compileGetArrayLength(Node* node)
+        speculationCheck(UnexpectedResizableArrayBufferView, JSValueSource::unboxedCell(baseGPR), node, branchTest8(NonZero, Address(baseGPR, JSArrayBufferView::offsetOfMode()), TrustedImm32(isResizableOrGrowableSharedMode)));
+#if USE(LARGE_TYPED_ARRAYS)
+    load64(Address(baseGPR, JSArrayBufferView::offsetOfLength()), resultGPR);
+    speculationCheck(ExitKind::Overflow, JSValueSource(), nullptr, branch64(Above, resultGPR, TrustedImm64(std::numeric_limits<int32_t>::max())));
```
Let's omit `ExitKind::` to align it to the other code.

hyjorc1 replied:
There is a conflict of `ExitKind::Overflow` and `ResultCondition::Overflow` in the DFGSpeculativeJIT.cpp.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -8132,6 +8132,46 @@ void SpeculativeJIT::compileGetArrayLength(Node* node)
+        breakpoint();
+        doneCases.link(this);
+
+        speculationCheck(ExitKind::Overflow, JSValueSource(), nullptr, branch64(Above, resultGPR, TrustedImm64(std::numeric_limits<int32_t>::max())));
```
Let's omit ExitKind:: to align it to the other code.

hyjorc1 replied:
There is a conflict of `ExitKind::Overflow` and `ResultCondition::Overflow` in the DFGSpeculativeJIT.cpp.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -8132,6 +8132,46 @@ void SpeculativeJIT::compileGetArrayLength(Node* node)
+        outOfBounds.link(this);
+        flushRegisters();
+        callOperation(operationThrowTypeErrorFromDataViewByteLength, LinkableConstant::globalObject(*this, node));
+        breakpoint();
```
Jumping over `flushRegister` is not correct (`doneCases`). Let's just do OSR exit here.
```
speculationCheck(OutOfBounds, JSValueSource::unboxedCell(baseGPR), node, outOfBounds);
doneCases.link(this);
```

hyjorc1 replied:
Thanks, updated. Something learned today. If my understand it correctly here, we pass the slowValue (JSValueSource) when the failure involves a specific JS value that needs to be tracked/recovered. And we need to pass the highValue (for ValueProfile) when tracking a value profile for optimization (e.g., gathering type feedback). But when to pass `node`/`node->child1().node()` for the highValue we need check that: 
1. If the failure is tied to the operation itself (e.g., an arithmetic overflow or OOB), then pass `node`.
2. If the base object is the cause of failure, and the baseline JIT may need to re-evaluate it, then pass the base node.

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -4376,6 +4376,33 @@ bool ByteCodeParser::handleIntrinsicGetter(Operand result, SpeculatedType predic
 
     switch (variant.intrinsic()) {
+    case DataViewByteLengthIntrinsic: {
+        if (m_inlineStackTop->m_exitProfile.hasExitSite(m_currentIndex, BadIndexingType))
```
And let's also check `OutOfBounds` too.

Constellation commented on Source/JavaScriptCore/jit/JITOperations.h:
```diff
@@ -399,6 +399,8 @@ JSC_DECLARE_JIT_OPERATION(operationPutToScope, void, (JSGlobalObject*, const JSI
 JSC_DECLARE_JIT_OPERATION(operationReallocateButterflyToGrowPropertyStorage, char*, (VM*, JSObject*, size_t newSize));
 
+JSC_DECLARE_JIT_OPERATION(operationThrowTypeErrorFromDataViewByteLength, void, (JSGlobalObject* globalObject));
+
```
If we do OSR exit, probably we do not need this.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.h:
```diff
@@ -2137,7 +2139,8 @@ class AssemblyHelpers : public MacroAssembler {
 
     enum class TypedArrayField { Length, ByteLength };
-    void loadTypedArrayByteLengthImpl(GPRReg baseGPR, GPRReg valueGPR, GPRReg scratchGPR, GPRReg scratch2GPR, std::optional<TypedArrayType>, TypedArrayField);
+    std::pair<Jump, JumpList> loadTypedArrayByteLengthImpl(GPRReg baseGPR, GPRReg valueGPR, GPRReg scratchGPR, GPRReg scratch2GPR, std::optional<TypedArrayType>, TypedArrayField);
```
I would like to use `std::tuple` for newer code.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.h:
```diff
@@ -1210,10 +1210,12 @@ class AssemblyHelpers : public MacroAssembler {
 #else
     JumpList branchIfResizableOrGrowableSharedTypedArrayIsOutOfBounds(GPRReg, GPRReg, GPRReg, std::optional<TypedArrayType>) { return { }; }
     void loadTypedArrayByteLength(GPRReg, GPRReg, GPRReg, GPRReg, TypedArrayType) { }
+    std::pair<Jump, JumpList> loadDataViewByteLength(GPRReg, GPRReg, GPRReg, GPRReg, TypedArrayType) { return { }; };
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.h:
```diff
@@ -1210,10 +1210,12 @@ class AssemblyHelpers : public MacroAssembler {
 #if USE(JSVALUE64)
     JumpList branchIfResizableOrGrowableSharedTypedArrayIsOutOfBounds(GPRReg baseGPR, GPRReg scratchGPR, GPRReg scratch2GPR, std::optional<TypedArrayType>);
     void loadTypedArrayByteLength(GPRReg baseGPR, GPRReg valueGPR, GPRReg scratchGPR, GPRReg scratch2GPR, TypedArrayType);
+    std::pair<Jump, JumpList> loadDataViewByteLength(GPRReg baseGPR, GPRReg valueGPR, GPRReg scratchGPR, GPRReg scratch2GPR, TypedArrayType);
```
Ditto.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -5771,6 +5777,45 @@ IGNORE_CLANG_WARNINGS_END
+                });
+                return patchpoint;
+            }
+
```
If we use OutOfBounds OSR exit, probably the code can be significantly simpler.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -3145,6 +3145,40 @@ void SpeculativeJIT::compileNewTypedArrayWithInt52Size(Node* node)
+        GPRReg baseGPR = base.gpr();
+        GPRReg scratch1GPR = scratch1.gpr();
+        GPRReg resultGPR = result.gpr();
+
```
Need `speculateDataViewObject`.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -3145,6 +3145,40 @@ void SpeculativeJIT::compileNewTypedArrayWithInt52Size(Node* node)
+        outOfBounds.link(this);
+        flushRegisters();
+        callOperation(operationThrowTypeErrorFromDataViewByteLength, LinkableConstant::globalObject(*this, node));
+        breakpoint();
```
Let's use OutOfBounds OSR exit.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -3145,6 +3145,40 @@ void SpeculativeJIT::compileNewTypedArrayWithInt52Size(Node* node)
+    GPRTemporary result(this, Reuse, base);
+    GPRReg baseGPR = base.gpr();
+    GPRReg resultGPR = result.gpr();
+
```
Need `speculateDataViewObject`.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -5872,6 +5917,18 @@ IGNORE_CLANG_WARNINGS_END
 
+    void compileDataViewGetByteLength()
+    {
+        LValue length = typedArrayLength(lowCell(m_node->child1()), m_node->mayBeResizableOrGrowableSharedArrayBuffer(), std::optional { TypeDataView }, m_node->child1());
```
You need to use lowDataViewObject.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -5882,6 +5939,12 @@ IGNORE_CLANG_WARNINGS_BEGIN("missing-noreturn")
+    void compileDataViewGetByteLengthAsInt52()
+    {
+        RELEASE_ASSERT(sizeof(size_t) == sizeof(uint64_t));
+        setStrictInt52(typedArrayLength(lowCell(m_node->child1()), m_node->mayBeResizableOrGrowableSharedArrayBuffer(), { TypeDataView }, m_node->child1()));
```
You need to use lowDataViewObject.

Constellation commented on Source/JavaScriptCore/assembler/AbortReason.h:
```diff
@@ -56,6 +56,7 @@ enum AbortReason {
     DFGIneffectiveWatchpoint                          = 190,
     DFGNegativeStringLength                           = 200,
     DFGSlowPathGeneratorFellThrough                   = 210,
+    DFGUnexpectedDataViewException                    = 215,
```
Let's remove this since it is no longer used.

Constellation commented on without_with_Microbenchmarks_yijia-mbp_20250310_2345.json:
Remove this file

Constellation commented on without_with_Microbenchmarks_yijia-mbp_20250310_2345_report.txt:
Remove this file.

hyjorc1 committed:
[JSC] Optimize DataView.prototype.byteLength handling in JIT
https://bugs.webkit.org/show_bug.cgi?id=289365
rdar://146511506

Reviewed by Yusuke Suzuki.

Introduced DataViewByteLengthIntrinsic to improve inline caching and JIT
optimizations for DataView.prototype.byteLength. With this change, the
byteLength property’s order is rearranged in the DataView property table.
Consequently, object-expected.txt has been updated accordingly.

MicroBenchmarks Results:

                                           without                     with

dataview-byte-length-small-array       14.7764+-0.0405     ^      4.6604+-0.0182        ^ definitely 2.8164x faster

* JSTests/microbenchmarks/dataview-byte-length-small-array.js: Added.
(test):
* JSTests/stress/dataview-byte-length-large-array-oob.js: Added.
(test):
(i.i.testLoopCount.1.catch):
* JSTests/stress/dataview-byte-length-large-array.js: Added.
(test):
* JSTests/stress/dataview-byte-length-small-array-oob.js: Added.
(test):
(i.i.testLoopCount.1.catch):
* JSTests/stress/dataview-byte-length-small-array.js: Added.
(test):
* Source/JavaScriptCore/assembler/AbortReason.h:
* Source/JavaScriptCore/bytecode/GetByVariant.cpp:
(JSC::GetByVariant::canMergeIntrinsicStructures const):
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheCompiler::canEmitIntrinsicGetter):
(JSC::InlineCacheCompiler::emitIntrinsicGetter):
* Source/JavaScriptCore/bytecode/IntrinsicGetterAccessCase.cpp:
(JSC::IntrinsicGetterAccessCase::doesCalls const):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGArrayMode.cpp:
(JSC::DFG::ArrayMode::alreadyChecked const):
(JSC::DFG::toTypedArrayType):
(JSC::DFG::toArrayType):
* Source/JavaScriptCore/dfg/DFGArrayMode.h:
(JSC::DFG::ArrayMode::isSomeTypedArrayViewOrDataView const):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicGetter):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByVal):
(JSC::DFG::SpeculativeJIT::compileGetTypedArrayLengthAsInt52):
(JSC::DFG::SpeculativeJIT::compilePutByVal):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::typedArrayLength):
(JSC::FTL::DFG::LowerDFGToB3::compileGetArrayLength):
(JSC::FTL::DFG::LowerDFGToB3::compileGetTypedArrayLengthAsInt52):
(JSC::FTL::DFG::LowerDFGToB3::compileGetByValImpl):
(JSC::FTL::DFG::LowerDFGToB3::compilePutByVal):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::loadTypedArrayByteLengthImpl):
(JSC::AssemblyHelpers::loadTypedArrayByteLengthCommonImpl):
(JSC::AssemblyHelpers::loadTypedArrayByteLength):
(JSC::AssemblyHelpers::loadDataViewByteLength):
(JSC::AssemblyHelpers::loadTypedArrayLength):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::loadDataViewByteLength):
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/jit/JITOperations.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
(JSC::JSDataViewPrototype::create):
(JSC::JSDataViewPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSDataViewPrototype.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):

Canonical link: https://commits.webkit.org/291993@main
